// @sap/xb-msg/examples/cfg/multi-destination-produce.js
'use strict';

module.exports = {
    destinations : [
        {
            name: 'broker-01',
            type: 'amqp-v091',

            net: {
                host          : '127.0.0.1',
                port          : 5672
            },
            sasl: {
                user          : 'guest',
                password      : 'guest'
            },
            amqp: {
                vhost         : '/',
            },
            ostreams          : {
                MyOutA        : { channel : 1, exchange : 'amq.topic', routingKey : 'Cars.Velocity.kilometersPerHour', confirms : false }
            },
        },
        {
            name: 'broker-02',
            type: 'mqtt-v311',

            net: {
                host          : '127.0.0.1',
                port          : 1883
            },
            credentials: {
                user          : '',
                password      : ''
            },
            mqtt: {
                clientID      : '',
                keepAlive     : 60
            },
            ostreams          : {
                MyOutB        : { topic : 'Cars/Velocity/kilometersPerHour', qos : 0 }
            },
        }
    ]
};
