// @sap/xb-msg-amqp-v091/lib/methods/confirm.js
'use strict';

const { Method } = require('../serialize/types');
const constants = require('../serialize/constants');
const SELECT = constants.METHOD.CONFIRM.SELECT;
const SELECT_OK = constants.METHOD.CONFIRM.SELECT_OK;


/**
 * @struct
 * @private
 */
class ConfirmSelect extends Method {

    /**
     *
     */
    constructor(noWait = true) {
        super(SELECT);

        /**
         * @type {boolean}
         */
        this.noWait = noWait;
    }

    /**
     * @param {!Decoder} decoder
     * @override
     */
    deserialize(decoder) {
        this.noWait = decoder.readBitSet8Inverse(1)[0];
    }

    /**
     * @param {!Encoder} encoder
     * @override
     */
    serialize(encoder) {
        super.serialize(encoder);
        encoder.writeBitSet8Inverse(this.noWait);
    }

}

/**
 * @struct
 * @private
 */
class ConfirmSelectOk extends Method {

    constructor() {
        super(SELECT_OK, SELECT);
    }

    /**
     * @param {!Decoder} decoder
     * @override
     */
    deserialize(decoder) {
    }

    /**
     * @param {!Encoder} encoder
     * @override
     */
    serialize(encoder) {
    }

}

module.exports = {
    Select : ConfirmSelect,
    SelectOk : ConfirmSelectOk
};
