// @sap/xb-msg-mqtt-v311/lib/serialize/encoder.js
'use strict';

const { EC, ErrMsg } = require('../errors');
const constants = require('./constants');
const CONTINUATION_BIT = constants.PACKET.LENGTH.CONTINUATION_BIT;
const LEN_VALUE_BASE = constants.PACKET.LENGTH.VALUE_BASE;
const LEN_UPPER_LIMIT = constants.PACKET.LENGTH.UPPER_LIMIT;

/**
 * @callback ChunkAllocate
 * @param {number=} minSize
 * @return {!Buffer}
 */

/**
 * @callback ChunkDispatch
 * @param {!Buffer} useChunk
 * @param {?Buffer} rawChunk
 */

/**
 * @private
 */
class Encoder {

    /**
     * @param {!StreamParameters} options
     * @param {!ChunkAllocate} allocate
     * @param {!ChunkDispatch} dispatch
     */
    constructor(options, allocate, dispatch) {

        /**
         * @type {number}
         */
        this.statChunkCount = 0;

        /**
         * @type {number}
         */
        this.statFrameCount = 0;

        /**
         * @const {number}
         * @private
         */
        this._payloadCopyLimit = options.ostreamPayloadCopyLimit;

        /**
         * @type {?Buffer}
         * @private
         */
        this._chunkWriting = null;

        /**
         * @type {number}
         * @private
         */
        this._chunkRemaining = 0;

        /**
         * @type {number}
         * @private
         */
        this._chunkOffset = 0;

        /**
         * @type {number}
         * @private
         */
        this._chunkProduced = 0;

        /**
         * @type {number}
         * @private
         */
        this._frameProduced = 0;

        /**
         * @type {number}
         * @private
         */
        this._frameDeclared = 0;

        /**
         * @type {!ChunkAllocate}
         * @private
         */
        this._chunkAllocate = allocate;

        /**
         * @type {!ChunkDispatch}
         * @private
         */
        this._chunkDispatch = dispatch;
    }

    /**
     *
     */
    stop() {
        this.push = () => nop(false);
        this.flush = nop;
    }

    /**
     * @param {!Packet} packet
     * @return {boolean} - dirty
     */
    push(packet) {
        let length = packet.serializing();

        if (this._frameProduced)
            throw ErrMsg(EC.ENCODE_INTERNAL_ERROR);
        if (length >= LEN_UPPER_LIMIT)
            throw ErrMsg(EC.DECODE_REM_LEN_EXCEEDS, length, LEN_UPPER_LIMIT);

        this.writeUInt8(packet.type | packet.flags);
        do {
            let byte = length % LEN_VALUE_BASE;
            length = Math.trunc(length / LEN_VALUE_BASE);
            if (length) byte |= CONTINUATION_BIT;
            this.writeUInt8(byte);
        } while (length);
        packet.serialize(this);

        ++this.statFrameCount;
        this._frameProduced = 0;
        return this._chunkProduced > 0;
    }

    /**
     * @param {number} length
     */
    flush(length = 0) {
        if (this._chunkProduced) {
            ++this.statChunkCount;
            const useChunk = this._chunkRemaining ? this._chunkWriting.slice(0, this._chunkProduced) : this._chunkWriting;
            const rawChunk = this._chunkWriting;
            this._chunkDispatch(useChunk, rawChunk);
        } else if (length === 0) {
            return;
        }
        this._chunkWriting = this._chunkAllocate(length);
        this._chunkProduced = 0;
        this._chunkRemaining = this._chunkWriting.length;
        this._chunkOffset = 0;
    }

    /**
     * @param {number} length
     * @private
     */
    _writing(length) {
        if (length > this._chunkRemaining) this.flush(length);
        this._chunkOffset = this._chunkProduced;
        this._chunkProduced += length;
        this._chunkRemaining -= length;
        this._frameProduced += length;
    }

    /**
     * @param {!Buffer} value
     */
    writeDataBlob(value) {
        if (value.length < this._payloadCopyLimit) {
            this._writing(value.length);
            value.copy(this._chunkWriting, this._chunkOffset);
        } else {
            this.flush();
            this._chunkDispatch(value, null);
            this._frameProduced += value.length;
        }
    }

    /**
     * @param {!number} value
     * @throws {Error}
     * @final
     */
    writeUInt8(value) {
        this._writing(1);
        this._chunkWriting.writeUInt8(value, this._chunkOffset);
    }

    /**
     * @param {number} value
     * @throws {Error}
     * @final
     */
    writeUInt16(value) {
        this._writing(2);
        this._chunkWriting.writeUInt16BE(value, this._chunkOffset);
    }

    /**
     * @param {string} value
     * @throws {Error}
     * @final
     */
    writeString(value) {
        if (value && value.length) {
            this.writeUInt16(value.length);
            this._writing(value.length);
            this._chunkWriting.write(value, this._chunkOffset, value.length, 'utf8');
        } else {
            this.writeUInt16(0x0000);
        }
    }

    /**
     * @param {Buffer} value
     */
    writeBlob16(value) {
        if (value && value.length) {
            this.writeUInt16(value.length);
            this.writeDataBlob(value);
        } else {
            this.writeUInt16(0x0000);
        }
    }

}

/**
 * @param {*=} result
 * @return {*}
 * @private
 */
function nop(result) { return result; }

module.exports = {
    Encoder : Encoder
};
